[BEAM-65] Splittable DoFn: Gives the runner access to RestrictionTracker

jkff opened the pull request:
Changes the SplittableParDo transform so that ProcessFn uses a
runner-supplied hook to run the @ProcessElement method, giving
it, among other things, the RestrictionTracker, so the runner
can initiate checkpointing/splitting with it at will.

Introduces a default implementation of said hook, which limits
the number of outputs and duration of the call. This implementation
is used in tests and in Direct runner. Dataflow Streaming runner
will also use this implementation, while Dataflow Batch runner
will be more sophisticated.

R: @kennknowles

jkff commented:
Oops, forgot to push a file. Fixed.

jkff commented:
Actually hold on with reviewing this for a bit while I understand whether this is really the best way to abstract this, by looking at Dataflow worker.

jkff commented:
retest this please

jkff commented:
Definitely ready for review. I have a prototype in Dataflow runner doesn't run yet, but demonstrates that this is the right API.

jkff commented:
Ping.

kennknowles commented on runners/core-java/src/main/java/org/apache/beam/runners/core/OutputAndTimeBoundedSplittableProcessElementInvoker.java:
```diff
@@ -0,0 +1,231 @@
+      }
+    }
+
+    private synchronized void initiateCheckpoint() {
```
It seems the purpose here is to make sure only one thread ever enters the (presumably-expensive) act of `tracker.checkpoint()`.

To embed the access policy and use futures more idiomatically, I think you could give `checkpoint` the type `AtomicReference<Future<CheckpointT>>`.

Failing that, give it a getter and a setter that are both synchronized, put them next to each other, and call them out loudly on the field declaration.

jkff replied:
I couldn't come up with cleaner code for this, but I moved the accesses to checkpoint closer together, and added a bunch of comments.

jkff commented on runners/core-java/src/main/java/org/apache/beam/runners/core/OutputAndTimeBoundedSplittableProcessElementInvoker.java:
```diff
@@ -0,0 +1,231 @@
+    final ProcessContext processContext = new ProcessContext(element, tracker);
+    DoFn.ProcessContinuation cont =
+        invoker.invokeProcessElement(
+            new DoFnInvoker.FakeArgumentProvider<InputT, OutputT>() {
```
Done, put in all the methods explicitly. A separate PR removes the annoying inputProvider/outputReceiver methods.

jkff commented:
PTAL.

dhalperi commented:
What JIRA is this under? What's the rationale for this change?

kennknowles commented:
Yea this should be tagged with [BEAM-65](https://issues.apache.org/jira/browse/BEAM-65).

jkff commented:
@dhalperi Tagged with JIRA; the rationale for this change is, I believe, explained in the commit message.

dhalperi commented:
+1 thanks Eugene!

jkff commented:
Added tests (while I was at it, moved OffsetRange and its Tracker to SDK since that'd be the third time I'd use it, including some pending code of mine where I use it in an integration test).

jkff commented:
retest this please

jkff commented:
Coveralls is weird. It says SplittableProcessElementsEvaluatorFactory is basically not covered, but it is covered by SplittableParDo tests in direct runner.

kennknowles commented:
This is a Jacoco configuration issue. All coveralls does is present the data, which is essentially available. I won't block anything on Coveralls.

To easily browse yourself, run `mvn jacoco:report`

dhalperi commented:
@jkff -- I believe this is induced by our crazy test setup. The SDK tests are covered in the direct runner module, etc. The issue is that we have not asked jacoco to combine these things into one report.

I spent a few minutes trying to fix this using newer versions of plugins that claim to be able to do it, but stalled: https://github.com/apache/beam/pull/1800

kennknowles commented:
The Java code LGTM. Seems to have pulled in a ton of Python stuff.

kennknowles commented:
I'm going to get green, fire off a postcommit, then merge. If  you'd like to fix up to preserve merge coherence you can, or I will squash when I merge, losing the history.

jkff commented:
Squashing is fine.

kennknowles commented:
https://builds.apache.org/view/Beam/job/beam_PostCommit_Java_RunnableOnService_Dataflow/2167/
https://builds.apache.org/view/Beam/job/beam_PostCommit_Java_RunnableOnService_Flink/1506/
https://builds.apache.org/view/Beam/job/beam_PostCommit_Java_RunnableOnService_Spark/816/
https://builds.apache.org/view/Beam/job/beam_PostCommit_Java_RunnableOnService_Apex/372/
